### ISSUE TITLE
The relates_to field gets lost when events are decrypted.

### ISSUE THREAD

**Issue by poljar** [matrix-org/matrix-rust-sdk#64]

**Comment by DevinR528** [comment]
Is this a problem with https://github.com/ruma/ruma/blob/master/ruma-events/src/room/encrypted.rs#L55? Does it need a `relates_to` field?

**Comment by poljar** [comment]
Yeah, either that or a way to access unknown fields. I'm a bit wary on just adding the relates to field since realistically it should be encrypted, though that's still far away from reality.

Perhaps adding a way to access the unknown fields would be welcome anyways, to handle additional custom and unspecced data that might be added by clients.

@jplatte do we have something planned for Ruma with regards to the custom fields?

**Comment by jplatte** [comment]
I don't actually fully understand this issue. For custom and unspecced data, you can deserialize `EventJson` to arbitrary `Deserialize` structs in addition to or instead of the expected event type.

**Comment by poljar** [comment]
The `relates_to` field needs to sent out in the content unencrypted, so it will sit by the ciphertext of an encrypted content. This is sadly in spec limbo.

I guess de-serializing twice isn't that bad to fish the `relates_to` field out.

**Comment by DevinR528** [comment]
Would it be possible to add a `pre_spec/final_comment` feature that makes __no__ semVer promises but could be where all the stuff that will most likely go into the spec soon-ish lives? I do see the maintenance burden this would add to ruma, keeping track of it all would be a pain, but I wonder if the escape hatch would be worth it? Just a thought...

@jplatte

**Comment by jplatte** [comment]
@DevinR528 The issue with that is that matrix-rust-sdk needs to use a released version of ruma, otherwise it can't be released itself. We could potentially release a `ruma-next`... But for that we'd need `ruma-events-next` and `ruma-client-api-next` too probably. I guess if there are strong reasons for it, I'll consider it.

**Comment by DevinR528** [comment]
Wouldn't something like this work in the normal `master` release?

```rust
// ruma_events/src/room/encrypted.rs
pub struct MegolmV1AesSha2Content {
    /// The encrypted content of the event.
    pub ciphertext: String,

    /// The Curve25519 key of the sender.
    pub sender_key: String,

    /// The ID of the sending device.
    pub device_id: DeviceId,

    /// The ID of the session used to encrypt the message.
    pub session_id: String,

    #[cfg(feature = "pre_spec")]
    relates_to: Whatever,
}
```
```toml
# then used in dependant crates
ruma = {version="0.1.1", features = ["pre_spec"]}
```
Or does that break something I'm not seeing? The other slight advantage would be testing out future implementations of coming features/spec changes.

**Comment by jplatte** [comment]
Hm yeah, that could work 🤔

**Comment by poljar** [comment]
This has been fixed by 3f41e5071b0058cd7b73cd891de65d80ff0ea707 for the decryption and by 5babd71341fc5ec3b0d03e23a4f4c08e80e43fae for the encryption side.
